src/pudl/extract/phmsagas.py

```diff
@@ -0,0 +1,84 @@
+"""Retrieves data from PHMSA natural gas spreadsheets for analysis.
+
+This modules pulls data from PHMSA's published Excel spreadsheets.
+"""
+
+
+from dagster import AssetOut, Output, multi_asset
+
+import pudl.logging_helpers
+from pudl.extract import excel
+
+logger = pudl.logging_helpers.get_logger(__name__)
+
+
+class Extractor(excel.GenericExtractor):
+    """Extractor for the excel dataset PHMSA."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the module.
+
+        Args:
+            ds (:class:datastore.Datastore): Initialized datastore.
+        """
+        self.METADATA = excel.Metadata("phmsagas")
+        self.cols_added = []
+        super().__init__(*args, **kwargs)
+
+    def process_final_page(self, df, page):
+        """Drop columns that get mapped to other assets.
+
+        Older years of PHMSA data may have fewer pages and tables of data than newer
+        data. This means one page may get split into multiple tables by column. To
+        prevent each table from containing all columns from these older years, filter by
+        the list of columns specified for the page, with a warning.
+        """
+        to_drop = [
+            c
+            for c in df.columns
+            if c not in self._metadata.get_all_columns(page)
+            and c not in self.cols_added
+        ]
+        if to_drop:
+            logger.warning(
+                f"Dropping columns {to_drop} that are not mapped to this asset."
+            )
+            df = df.drop(columns=to_drop, errors="ignore")
+        return df
+
+
+# TODO (bendnorman): Add this information to the metadata
+raw_table_names = (
+    "raw_phmsagas__yearly_distribution",
+    "raw_phmsagas__yearly_transmission_gathering_summary_by_commodity",
+    "raw_phmsagas__yearly_miles_of_gathering_pipe_by_nps",
+    "raw_phmsagas__yearly_miles_of_transmission_pipe_by_nps",
+)
+
+phmsagas_raw_dfs = excel.raw_df_factory(Extractor, name="phmsagas")
+
+
+# # TODO (bendnorman): Figure out type hint for context keyword and multi_asset return
+@multi_asset(
+    outs={table_name: AssetOut() for table_name in sorted(raw_table_names)},
+    required_resource_keys={"datastore", "dataset_settings"},
+)
+def extract_phmsagas(context, phmsagas_raw_dfs):
```

Comment:
I'm not sure of a great alternative to `multi_assets` given our current setup. The issue with multi assets is they limit parallelism and don't make it easier to execute individual assets. We could have a factory that extracts a single page and concatenates all of the requested years. This way each page is a separate asset that can be executed on its own. The factory would probably create a `graph_asset` that handles some parallelism for extracting separate years. This might be tricky to implement right now because the `GenericExtractor` is currently built to extract all pages for a single year.

Maybe @zaneselvans or @dstansby have more thoughts on how we can rework the extractor to be a little more dagster friendly?

A small improvement you can make is to [enable multi asset subsetting](https://docs.dagster.io/concepts/assets/multi-assets#subsetting-multi-assets) so we can execute downstream assets in isolation. I started doing this in a stale PR  #2773.

Reply:
It might make more sense to not enable multi_asset subsetting for phmsa until #2773 is closed out. It's been prioritized because of other inframundo work.

Reply:
I agree, I'd rather just implement it all in one place. Helpful to know what the status of this is, though!